akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -1158,32 +1158,36 @@ private[stream] final class DropWithin[T](timeout: FiniteDuration) extends Simpl
 private[stream] final class Reduce[T](f: (T, T) ⇒ T) extends SimpleLinearGraphStage[T] {
   override def initialAttributes: Attributes = DefaultAttributes.reduce
 
-  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
-    override def toString = s"Reduce.Logic(aggregator=$aggregator)"
-    var aggregator: T = _
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler { self ⇒
+      override def toString = s"Reduce.Logic(aggregator=$aggregator)"
+      var aggregator: T = _
 
-    setHandler(in, new InHandler {
-      override def onPush(): Unit = {
-        aggregator = grab(in)
-        pull(in)
-        setHandler(in, rest)
-      }
-    })
-    def rest = new InHandler {
       override def onPush(): Unit = {
         aggregator = f(aggregator, grab(in))
         pull(in)
       }
+
       override def onUpstreamFinish(): Unit = {
         push(out, aggregator)
         completeStage()
       }
-    }
 
-    setHandler(out, new OutHandler {
       override def onPull(): Unit = pull(in)
-    })
-  }
+
+      setHandler(in, new InHandler { // Initial input handler
+        override def onPush(): Unit = {
+          aggregator = grab(in)
+          pull(in)
+          setHandler(in, self)
+        }
+
+        override def onUpstreamFinish(): Unit =
+          failStage(new UnsupportedOperationException("empty stream reduce"))
+      })
+
+      setHandler(out, self)
```

Comment:
I was a bit iffy about using self here but it actually looks good and should be worth it, thanks :) 
As suggested in the other PR I'll copy that optimisation :+1: